File: fetch_valid_words.py
import requests
from bs4 import BeautifulSoup
import time
import os
import sys

def fetch_words_from_page(page_number):
    url = f'https://www.bestwordlist.com/5letterwords{"page" + str(page_number) + ".htm" if page_number > 1 else ".htm"}'
    response = requests.get(url)
    response.raise_for_status()
    
    soup = BeautifulSoup(response.text, 'html.parser')
    
    words = set()
    spans = soup.find_all('span', class_=['mt', 'rd'])
    
    for span in spans:
        text = span.get_text()
        words_list = text.split()
        for word in words_list:
            if len(word) == 5:
                words.add(word.lower())
    
    return words

def main():
    all_words = set()
    script_dir = os.path.dirname(os.path.abspath(__file__))
    output_file = os.path.join(script_dir, 'valid_words.txt')
    
    total_pages = 26
    
    for page_number in range(1, total_pages + 1):
        sys.stdout.write(f"\rFetching Page {page_number} of {total_pages}: [{'#' * (page_number * 50 // total_pages)}{'.' * (50 - (page_number * 50 // total_pages))}]")
        sys.stdout.flush()
        words = fetch_words_from_page(page_number)
        all_words.update(words)
        time.sleep(1)
    
    sys.stdout.write("\n")
    
    with open(output_file, 'w') as f:
        for word in sorted(all_words):
            f.write(f"{word}\n")
    
    print(f"Finished fetching words and saved to {output_file}")

if __name__ == "__main__":
    main()

File: utils/fetch_valid_words.cpp
#include <iostream>
#include <fstream>
#include <set>
#include <string>
#include <curl/curl.h>
#include <gumbo.h>
#include <unistd.h>

static size_t WriteCallback(void* contents, size_t size, size_t nmemb, std::string* s) {
    size_t newLength = size * nmemb;
    try {
        s->append((char*)contents, newLength);
    } catch (std::bad_alloc& e) {
        return 0;
    }
    return newLength;
}

std::string fetch_page_content(int page_number) {
    std::string url = "https://www.bestwordlist.com/5letterwords";
    if (page_number > 1) {
        url += "page" + std::to_string(page_number) + ".htm";
    } else {
        url += ".htm";
    }

    CURL* curl;
    CURLcode res;
    std::string readBuffer;

    curl = curl_easy_init();
    if (curl) {
        curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteCallback);
        curl_easy_setopt(curl, CURLOPT_WRITEDATA, &readBuffer);
        res = curl_easy_perform(curl);
        curl_easy_cleanup(curl);
    }

    return readBuffer;
}

void search_for_words(GumboNode* node, std::set<std::string>& words) {
    if (node->type != GUMBO_NODE_ELEMENT) return;

    GumboAttribute* class_attr;
    if ((node->v.element.tag == GUMBO_TAG_SPAN) && 
        (class_attr = gumbo_get_attribute(&node->v.element.attributes, "class")) &&
        (strcmp(class_attr->value, "mt") == 0 || strcmp(class_attr->value, "rd") == 0)) {

        GumboNode* text = static_cast<GumboNode*>(node->v.element.children.data[0]);
        if (text->type == GUMBO_NODE_TEXT) {
            std::string content = text->v.text.text;
            size_t pos = 0;
            while ((pos = content.find(' ', pos)) != std::string::npos) {
                std::string word = content.substr(0, pos);
                if (word.size() == 5) {
                    words.insert(word);
                }
                content = content.substr(pos + 1);
            }
            if (content.size() == 5) {
                words.insert(content);
            }
        }
    }

    GumboVector* children = &node->v.element.children;
    for (unsigned int i = 0; i < children->length; ++i) {
        search_for_words(static_cast<GumboNode*>(children->data[i]), words);
    }
}

std::set<std::string> fetch_words_from_page(int page_number) {
    std::string page_content = fetch_page_content(page_number);

    GumboOutput* output = gumbo_parse(page_content.c_str());
    std::set<std::string> words;
    search_for_words(output->root, words);
    gumbo_destroy_output(&kGumboDefaultOptions, output);

    return words;
}

int main() {
    std::set<std::string> all_words;
    std::string output_file = "../data/valid_words.txt";
    int total_pages = 26;

    for (int page_number = 1; page_number <= total_pages; ++page_number) {
        std::cout << "\rFetching Page " << page_number << " of " << total_pages << ": ["
                  << std::string(page_number * 50 / total_pages, '#')
                  << std::string(50 - (page_number * 50 / total_pages), '.') << "]";
        std::cout.flush();

        std::set<std::string> words = fetch_words_from_page(page_number);
        all_words.insert(words.begin(), words.end());
        sleep(1);
    }

    std::cout << std::endl;

    std::ofstream out(output_file);
    for (const auto& word : all_words) {
        out << word << std::endl;
    }

    std::cout << "Finished fetching words and saved to " << output_file << std::endl;

    return 0;
}
